### DISCUSSION TITLE
Stat calculations for RBY/GSC

### DISCUSSION THREAD

**Discussion by ShiraTheMogul** [GregLyons/poke-team#3] [Ideas]
Heya, I saw your thread on Smogon and really like the aesthetic of this tool. There are a few things I'm interested in that pertain to RBY and GSC, as the stat calculation there is a bit different to what you may be used to.

### Stat Exp
Stat Exp is RBY and GSC's equivalent to EVs. They amount to the same thing, but the calculation is different; specifically, it's a range of 0-65535, not 0-252. RBYers are particularly attached to using the old notation, just a cultural thing community-wise. 

You can find the formulas for this [here](https://bulbapedia.bulbagarden.net/wiki/Stat#Generations_I_and_II), it would be incredible if there was either a toggle to view it this way and/or view it by default for these generations.

### Auto-setting the HP & Special DVs
The HP DV is set automatically and thus does not need to be done manually in your builder. Specifically, if the DV used is odd, it will add a specific number for each. 8 for Atk, 4 for Def, 2 for Spc, 1 for Spe. 

The SpA and SpD DV is shared in GSC to maintain compatability with RBY, so this should also be merged in some way. 

### Auto-setting Shininess in GSC
Shiny Pokemon happen under the following conditions;
- Attack DV must be 2, 3, 6, 7, 10, 11, 14, or 15
- Def, Spc, and Spe DVs must be 10
So when Shiny is ticked, you should set Attack to 15 and the rest to 10, wherein the HP DV sets itself.

### Auto-setting gender in GSC
[This article](https://www.smogon.com/ingame/guides/rby_gsc_stats#gender) sums it up tidily.

### QoL for GSC Marowak
If Marowak holds a Thick Club with max Attack, its Attack will overflow after using Swords Dance. Players usually set the Attack DV to 13 to alleviate this, as it'll reach 996 at +2. It would be nice to have this set itself upon adding a Thick Club so someone doesn't forget.

That's about all I have right now, this is a really nice tool and having these in would make it all-the-more hospitable to old gen players!

**Comment by GregLyons** [comment]
Glad to hear that you're enjoying it! :)

It's good to hear from someone who plays the older gens, as I have very little competitive experience with them. It sounds silly but I realized I just had the ability to change EVs completely disabled for Gens 1 and 2 (I saw in PokemonShowdown they have all the EVs set to 252, and I didn't think to check that you actually *can* modify the EVs still... idk 🤦‍♂️). So I'm definitely going to turn that option on today... (**EDIT**: that change should be live now)

Here's a high-level description of what I'm thinking: 

- Code in certain relationships (HP DV being determined by other DVs, and not able to be changed on its own).
- Add two buttons next to the Gen Filter, just like the Sw/Sh and BD/SP modes, but only present in Gens 1 and 2.
- First button is something like "STATEXP", which is the toggle you're talking about. I'll have that turned OFF by default, but you can just click it to turn it on.
- Second button is something like "AUTO" (for Gen 2 only), which will turn on a lot of the behavior you talk about (shiny, gender, Marowak). It'll be ON by default, but the user can turn it off if they like.

I have a couple questions for you on how best to implement these, which you can answer when you get the chance.

**Stat Exp**

I believe I can tweak the UI in the Teambuilder to account for the extra space that the larger numbers would require. Below is a screenshot on a kind of "minimum" screen size (after which the user needs to see everything--I want to change things so that everything fits in this minimum screen size). The upper-left is where the buttons go, the bottom-center is where I'll tweak the UI in StatExp mode (move Hidden Power and Nature over to the right give more vertical room to EVs; Nature is turned off prior to gen 3 anyway).

![statexp-changes](https://user-images.githubusercontent.com/89315619/163379294-77654922-3537-486b-a6f6-20d907162ab0.png)

What I want your opinion on is how to handle the EV sliders. It looks like StatExp doesn't have a *linear* relationship with stats, but instead a *square root* relationship. I have a linear step size of 4 for the EV sliders, since one wouldn't need to have an EV that isn't a multiple of 4. I'm thinking the most I could do is just use some arbitrary linear step size, and then rely on users explicitly entering their desired StatExp value instead. (Since the sliders and the text input are so connected, I can't just have the 0-252 system underlying the slider and display the StatExp 0-65535 number in the text input.)

Alternatively, I could just put the 0-65535 StatExp numbers on the right-hand side of the sliders, so that the user can *look* at the 0-65535 number changing as they drag the stat slider, but the slider (and hence the text input) still operates on a 0-252 system. Maybe that would feel more natural? Is the notation just an aesthetic thing, or do some people prefer using the *square root* relationship to the *linear* one when changing the stats? If it's the former, then I think I'd go with this alternative way.

**Hard-coded relationships**

*Special EVs/DVs*

I believe the special DVs are already shared for both Gens 1 and 2 in the app. By the gen 1 compatibility logic, I'm assuming special StatExp is merged as well?

*HP DVs*

I'll remove the ability to change the HP DV in Gens 1 and 2, since it's determined by the other DVs. It'll change according to those.

**Other auto-setting**

I'll work on coding in the shininess/gender restrictions. I'm thinking the easiest thing to do would be to use the "AUTO" toggle described above: when it's on and the user changes a DV to something invalid, the app will set the DVs to a valid value. Something like:

- **Shiny**: Two things:
  - If Def/Spc/Spe are not equal to 10, set equal to 10 after a brief delay
  - If Atk DV is an invalid value, set to next highest valid value (1 -> 2, 4 and 5 -> 6, etc.)
- **Gender**: If Atk DV goes above the valid maximum, reset to maximum after a brief delay (for 1:3 ratio, anything above 4 is set to 4, for 1:7 anything above 1 is set to 1, etc.).
- **Marowak**: If user selects Marowak in Gen 2 with Thick Club in Swords Dance, then automatically set Attack DV to 13. The user can then change the Attack DV after that to whatever they like (even >13; I believe just setting the DV to 13 at first will jog the user's memory about the overflow mechanic). 

Turning "AUTO" off will disable these Shiny/Gender/Marowak mechanics, and it'll be up to the user to make sure they have valid DVs. Thankfully I'm already using the gender ratio in the app to determine whether Pokemon are male only/female only/genderless, so I don't need to get extra data to implement the gender mechanic.



Let me know if there's an implementation for any of the above that would be more ergonomic for RBY/GSC users. I'll try to get all these things in by the end of next week, but once I get a couple done this week I'll merge those into `main` so that y'all can have access to them.

**Comment by ShiraTheMogul** [comment]
> Alternatively, I could just put the 0-65535 StatExp numbers on the right-hand side of the sliders, so that the user can look at the 0-65535 number changing as they drag the stat slider, but the slider (and hence the text input) still operates on a 0-252 system. Maybe that would feel more natural? Is the notation just an aesthetic thing, or do some people prefer using the square root relationship to the linear one when changing the stats? If it's the former, then I think I'd go with this alternative way.
It is mostly aesthetic, yeah. I think this route feels best, I don't think many would notice the unnatural-ness when using the slider. We usually hard-punch the numbers in when using save editors, etc, as we find them through math. 

> I believe the special DVs are already shared for both Gens 1 and 2 in the app. By the gen 1 compatibility logic, I'm assuming special StatExp is merged as well?
Yep!

I like what you're putting out here, looks great!

**Reply by GregLyons** [reply]
Here's what StatExp looks like next to the sliders. Is this what you meant by displaying the StatExp numbers? This is what I was thinking. I used the formula `EV * EV` to get them, so intermediate StatExp numbers will be inaccessible, but all possible relevant values should be (just like with EVs). Does that matter for competitive play?

![statexp-displayed](https://user-images.githubusercontent.com/89315619/163417302-8eb17691-39a1-4612-a79f-f5b7ca183528.png)
![statexp-versus](https://user-images.githubusercontent.com/89315619/163417304-ebba94e4-15b1-449c-aaeb-9e0f702453da.png)

I'm not so sure about putting the StatExp values in the 3x2 table (in the Teambuilder UI itself), as even with extra vertical space it might end up looking cluttered on smaller screens (e.g. laptop). If it's important I could tinker with it later to try and make it work, too.

Adding the StatExp to the sliders, on the other hand, is pretty trivial. If that suffices, I don't think it warrants a "STATEXP" toggle like I was thinking, as it's not really intrusive at all next to the sliders.

I'll get to the other stuff later, but wanted to share this now.

**EDIT**: Actually I should probably just let you test out how the StatExp feels. Since it's relatively unobtrusive, I just made it live. Let me know what you think

**Comment by GregLyons** [comment]
OK, I believe I've implemented this correctly. Here's an overview of the simpler changes.

- StatExp is displayed next to EV sliders in Gens 1 and 2. It's computed as `EV * EV`. Please let me know if there's some other way I should calculate it that's more in-line with how RBY/GSC players see/use it, as the main point is to give them something they're more accustomed to.
- HP DV slider is removed in Gens 1 and 2, and it is computed based on the other DVs.
- The SpA and SpD StatExp/DV are merged in Gens 1 and 2; specifically, changing one changes the other to match.
  - In Gen 1 only, the SpA/SpD StatExp/DV sliders are displayed as a single "Spc" (Special), as they are literally the same stat in Gen 1. (Still displayed as SpA/SpD in the summary window/stat graph.)
  - In Gen 2, the SpA/SpD StatExp/DV sliders are displayed separately. Let me know if I should change it to "Spc" as well.
- In Gen 2, when the Pokemon is shiny, Def/SpA/SpD/Spe cannot be changed from 10.
- Selecting both Thick Club AND Swords Dance on Gen 2 Marowak will set its Attack DV to 13. After this initial setting, one can modify the DVs afterwards to the desired value (even possibly 14/15, to get overflow).

As for "Auto-setting Shininess in GSC" and "Auto-setting gender in GSC", I came up with the following solution to ensure valid attack DVs. Essentially, I just determine the maximum valid attack DV given the gender/shiny values and set it to that. If that's not possible (i.e. 7:1 F:M ratio, female, AND shiny), then the shiny/gender value (the one not being directly modified) is set to either No/Male so that a possible attack DV can be found. All of these apply in Gen 2 only. Here's a more in-depth description:

- Setting the Pokemon's gender to female will automatically set its attack DV to the maximum valid attack DV, determined by its female:male gender ratio and its shiny value. 
  - If the Pokemon is not shiny, this is just a value equal to 1, 4, 7, 11, or 15.
  - If the Pokemon is shiny, then 7, 11, and 15 are still valid attack DVs. In the case of 4, the highest valid DV is 3. In the case of 1, there is no valid DV (as a shiny female Pokemon with 7:1 ratio cannot exist), so the shiny value is set to false to allow a valid attack DV, and the attack DV is set to 1.
-  Setting the Pokemon's shiny value to true will automatically set its attack DV to the maximum valid attack DV, again determined by its female:male gender ratio.
  - If the Pokemon is male or genderless, this value is 15.
  - Otherwise (Pokemon is female), it's either 3, 7, 11, or 15 as described above. In the case of 7:1 ratio, the gender is set to male to allow a valid attack DV (of 1).

Finally, when modifying the Attack DV directly (again in Gen 2) in the Teambuilder or in the Versus section, closing the slider menu (clicking "DONE" or otherwise closing the slider menu in the Teambuilder, or closing the popup in the Versus section), the Attack DV is set to the maximum valid Attack DV given the gender and shiny value. I decided this would be the best way to handle the Attack DV, as trying to reset the slider while the user is still working with it, or only allowing certain values on the slider, leads to janky behavior.


I believe that should handle all the edge cases, while nudging users towards valid Attack DVs in Gen 2. I assume that most RBY/GSC players are at least somewhat familiar with these mechanics, so they should know what's going on when the Attack DV is set indirectly (as opposed to directly by them). Thank you again, @PlagueVonKarma. I hope these changes make this app friendlier to Gen 1/2 players! If not, please let me know what feels off, and I can try to implement these features differently.
